README
************************************************************************
*               gpc++ - The Genetic Programming Kernel                 *
*                         C++ Class Library                            *
************************************************************************

What is the GP kernel?
======================

The GP kernel is a C++ class library that can be used to apply genetic
programming techniques to all kinds of problems.  The library defines
a class hierarchy.  An integral component is the ability to produce
automatically defined functions as found in Koza's "Genetic
Programming II".

Technical documentation in postscript format can be found in the "doc"
directory.  There is also a short introduction into genetic
programming (by Adam Fraser).

This system includes everything you ever wanted from a genetic
programming system but were always afraid to ask for, such as....

- Automatically defined functions (ADFs).

- Tournament and fitness proportionate selection.

- Demetic grouping

- Optional steady state genetic programming kernel (like the Adam
Fraser code).  Steady state means that not a complete new population
is built up every generation, but bad performing GPs are replaced by
newly evolved ones.  This uses approximately half the memory!

- Subtree crossover.

- Swap and shrink mutation.

- A way of changing every parameter of the system without
recompilation.

- Capacity for multiple populations (if you have the resources).

- Loading and saving of populations and genetic programs.

- Standard random number generator for complete portability.  This has
been improved: population sizes of over 250000 are possible now.

- Explanations of how to write the modules for evolving code.

- Examples.

- And the best thing for the experts: A parameter study about the
influence of every important parameter.

- Completely ANSI C++ compliant.

- Full and wonderful comments dispersed throughout the code.

- Internal parameter checks (can be turned off by a compiler switch).
Even if the programmer makes a mistake, the kernel complains.



This C++-code makes not use of templates and exceptions, so no C++
compatibility problems should arise.  

This software was compiled with the following compilers: 

- GNU C++-Compiler 2.7.0
- C++-Compiler by SUNSoft

Read the file INSTALL for how to compile the code.



LICENCE
=======

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 1, or (at your option)
any later version.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.


Copyright 1993, 1994 Adam P. Fraser and 1996, 1997 Thomas Weinbrenner

For comments, improvements, additions (or even money) contact:

Thomas Weinbrenner
Grauensteinstr. 26
35789 Laimbach
Germany
E-mail: thomasw@emk.e-technik.th-darmstadt.de
WWW:    http://www.emk.e-technik.th-darmstadt/~thomasw

  or 

(Address may be out of date)
Adam Fraser, Postgraduate Section, Dept of Elec & Elec Eng,
Maxwell Building, University Of Salford, Salford, M5 4WT, United Kingdom.
E-mail: a.fraser@eee.salford.ac.uk
Tel:    (UK) 061 745 5000 x3633
Fax:    (UK) 061 745 5999



History
************************************************************************
*               gpc++ - The Genetic Programming Kernel                 *
*                         C++ Class Library                            *
************************************************************************

0.50  (15/1/97)
---------------

First version

0.5.1 (28/1/97)
---------------

Shrink mutation faster

Bug in GP::load(): Length and depth was not calculated

config.h renamed to gpconfig.h

Prefix GP added to all names of module config

Bug in GPGene::chooseFunctionNode(): If no function node, floating
point exception

Bug in GP::mutate(): flag fitnessValid not set to 0

Steady State feature added.  The following had to change because of
this:

- Arguments and return value of GPPopulation::tournamentSelection(),
  GPPopulation::probabilisticSelection() changed.

- New function GPPopulation::selectIndices().

- New GPVariables parameter SteadyState.

New GPVariables parameter AddBestToNewPopulation, CrossoverProbability
and CreationProbability.  The Parameters ReproductionRation and
CreationRatio are no longer valid.

Selection mechanism for Demetic Migration is now one of the usual
ones, not just random selection.  (Was possible because of the new
function GPPopulation::selectIndices())

0.5.2 (13/2/97)
---------------

Bug in GPPopulation::calculateStatistics(): The calculation of the
worst of the population was wrong, and in some weird cases the best of
the population was wrong, too.  This was not a severe bug, but can
cause unnecessary trouble.

Bug in GP::compare: Returned value 1 when the trees were the same.

Documentation: Parameter study added.


src/gpv.cc

// gpv(ariables).cc

/* -------------------------------------------------------------------

gpc++ - The Genetic Programming Kernel

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 1, or (at your option)
any later version.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.


Copyright 1993, 1994 Adam P. Fraser and 1996, 1997 Thomas Weinbrenner

For comments, improvements, additions (or even money) contact:

Thomas Weinbrenner
Grauensteinstr. 26
35789 Laimbach
Germany
E-mail: thomasw@emk.e-technik.th-darmstadt.de
WWW:    http://www.emk.e-technik.th-darmstadt/~thomasw

  or 

(Address may be out of date)
Adam Fraser, Postgraduate Section, Dept of Elec & Elec Eng,
Maxwell Building, University Of Salford, Salford, M5 4WT, United Kingdom.
E-mail: a.fraser@eee.salford.ac.uk
Tel:    (UK) 061 745 5000 x3633
Fax:    (UK) 061 745 5999

------------------------------------------------------------------- */

// A nice simple class to contain all the genetic programming
// variables



#include <stdlib.h>
#include <fstream>
#include "gp.h"

using namespace std;

// Setup default values
GPVariables::GPVariables ()
{
  PopulationSize=100;
  NumberOfGenerations=20;
  CrossoverProbability=95.0;
  CreationProbability=2.0;
  CreationType=2;
  MaximumDepthForCreation=6;
  MaximumDepthForCrossover=17;
  SelectionType=GPTournamentSelection;
  TournamentSize=10;
  DemeticGrouping=0;
  DemeSize=100;
  DemeticMigProbability=100.0;
  SwapMutationProbability=0.0;
  ShrinkMutationProbability=0.0;
  AddBestToNewPopulation=1;
  SteadyState=1;
}



GPVariables::GPVariables (const GPVariables& gpo) : GPObject(gpo) 
{
  PopulationSize=gpo.PopulationSize;
  NumberOfGenerations=gpo.NumberOfGenerations;
  CrossoverProbability=gpo.CrossoverProbability;
  CreationProbability=gpo.CreationProbability;
  CreationType=gpo.CreationType;
  MaximumDepthForCreation=gpo.MaximumDepthForCreation;
  MaximumDepthForCrossover=gpo.MaximumDepthForCrossover;
  SelectionType=gpo.SelectionType;
  TournamentSize=gpo.TournamentSize;
  DemeticGrouping=gpo.DemeticGrouping;
  DemeSize=gpo.DemeSize;
  DemeticMigProbability=gpo.DemeticMigProbability;
  SwapMutationProbability=gpo.SwapMutationProbability;
  ShrinkMutationProbability=gpo.ShrinkMutationProbability;
  AddBestToNewPopulation=gpo.AddBestToNewPopulation;
  SteadyState=gpo.SteadyState;
}



// Write variables to ostream
void GPVariables::printOn (ostream& os)
{
  os << "PopulationSize            = " << PopulationSize
     << "\nNumberOfGenerations       = " << NumberOfGenerations
     << "\nCrossoverProbability      = " << CrossoverProbability
     << "\nCreationProbability       = " << CreationProbability
     << "\nCreationType              = ";
  switch (CreationType)
    {
    case GPVariable:
      os << "Variable";
      break;
    case GPGrow:
      os << "Grow";
      break;
    case GPRampedHalf:
      os << "Ramped Half and Half";
      break;
    case GPRampedVariable:
      os << "Ramped Variable";
      break;
    case GPRampedGrow:
      os << "Ramped Grow";
      break;
    case GPUserDefinedCreation:
      os << "User defined";
      break;
    default:
      GPExitSystem ("GPVariables::printOn", "Wrong creation type");
    }
  os << "\nMaximumDepthForCreation   = " << MaximumDepthForCreation
     << "\nMaximumDepthForCrossover  = " << MaximumDepthForCrossover
     << "\nSelectionType             = ";
  switch (SelectionType)
    {
    case GPProbabilisticSelection:
      os << "Probablistic selection";
      break;
    case GPTournamentSelection:
      os << "Tournament selection";
      break;
    case GPUserDefinedSelection:
      os << "User defined";
      break;
    default:
      GPExitSystem ("GPVariables::printOn", "Wrong selection type");
    }
  os << "\nTournamentSize            = " << TournamentSize
     << "\nDemeticGrouping           = " << (DemeticGrouping?"On":"Off")
     << "\nDemeSize                  = " << DemeSize
     << "\nDemeticMigProbability     = " << DemeticMigProbability
     << "\nSwapMutationProbability   = " << SwapMutationProbability
     << "\nShrinkMutationProbability = " << ShrinkMutationProbability
     << "\nAddBestToNewPopulation    = " << AddBestToNewPopulation
     << "\nSteadyState               = " << SteadyState
     << endl;
}



// Load operation
char* GPVariables::load (istream& is)
{
  is >> PopulationSize;
  is >> NumberOfGenerations;
  is >> CrossoverProbability;
  is >> CreationProbability;
  is >> CreationType;
  is >> MaximumDepthForCreation;
  is >> MaximumDepthForCrossover;
  is >> SelectionType;
  is >> TournamentSize;
  is >> DemeticGrouping;
  is >> DemeSize;
  is >> DemeticMigProbability;
  is >> SwapMutationProbability;
  is >> ShrinkMutationProbability;
  is >> AddBestToNewPopulation;
  is >> SteadyState;
  return NULL;
}



// Save operation
void GPVariables::save (ostream& os)
{
  os << PopulationSize << ' ';
  os << NumberOfGenerations << ' ';
  os << CrossoverProbability << ' ';
  os << CreationProbability << ' ';
  os << CreationType << ' ';
  os << MaximumDepthForCreation << ' ';
  os << MaximumDepthForCrossover << ' ';
  os << SelectionType << ' ';
  os << TournamentSize << ' ';
  os << DemeticGrouping << ' ';
  os << DemeSize << ' ';
  os << DemeticMigProbability << ' ';
  os << SwapMutationProbability << ' ';
  os << ShrinkMutationProbability << ' ';
  os << AddBestToNewPopulation << ' ';
  os << SteadyState << ' ';
}
